### ISSUE TITLE
E+ 8.9.0 doesn't recognize PlantEquipmentOperation:XXXX

### ISSUE THREAD

**Issue by jmarrec** [NREL/EnergyPlus#6584]
Issue overview
--------------

File used to run fine in 8.8.0, fails with a fatal error in the latest release 8.9.0.

```
Program Version,EnergyPlus, Version 8.9.0-40101eaafd, YMD=2018.03.26 10:49,
   ** Severe  ** <root>[PlantEquipmentOperationSchemes][Chilled Water Loop Operation Schemes][control_scheme_1_object_type] - "PlantEquipmentOperation:OutdoorDewpointDifference" - Failed to match against any enum values.
   **  Fatal  ** Errors occurred on processing input file. Preceding condition(s) cause termination.
   ...Summary of Errors that led to program termination:
   ..... Reference severe error count=1
   ..... Last severe error=<root>[PlantEquipmentOperationSchemes][Chilled Water Loop Operation Schemes][control_scheme_1_object_type] - "PlantEquipmentOperation:OutdoorDewpointDifference" - Failed to match against any enum values.
   ************* Warning:  Node connection errors not checked - most system input has not been read (see previous warning).
   ************* Fatal error -- final processing.  Program exited before simulations began.  See previous error messages.
   ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Terminated--Fatal Error Detected. 0 Warning; 1 Severe Errors; Elapsed Time=00hr 00min  0.32sec
```

Here is the defect file: [plant_op_deltatemp_schemes.rb.idf.txt](https://github.com/NREL/EnergyPlus/files/1847386/plant_op_deltatemp_schemes.rb.idf.txt)

I have another one that doesn't recognize `PlantEquipmentOperation:OutdoorWetBulb`: 
[plant_op_schemes.rb.idf.txt](https://github.com/NREL/EnergyPlus/files/1847396/plant_op_schemes.rb.idf.txt)

Another that doesn't recognize `PlantEquipmentOperation:OutdoorRelativeHumidity`: 
[plant_op_temp_schemes.rb.idf.txt](https://github.com/NREL/EnergyPlus/files/1847397/plant_op_temp_schemes.rb.idf.txt)

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): Ubuntu 16.04 LTS
 - Version of EnergyPlus (if using an intermediate build, include SHA): Version 8.9.0 Official (40101eaafd)

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by mjwitte** [comment]
@jmarrec According to the idd and the Input Output References, the file should not have run in v8.8.0. According to the documentation, some operation schemes apply only to PlantEquipmentOperation and some schemes apply only to CondenserEquipmentOperation. Input processing in v8.9.0 is more strict so the file fails.

@Myoldmopar A quick inspection of the code appears to show that all operation scheme types are valid for both PlantEquipmentOperation and CondenserEquipmentOperation, correct? But the IDD and I/O Ref were never updated to reflect this. Yet, the code only allows 

I do not see a workaround without recompiling the code, because editing the IDD will not alter the embedded JSON schema. idf workarounds fail because PlantLoop won't recognize a CondenserEquipmentOperationScheme name, and CondenserLoops don't allow boilers and chillers.

**Comment by jmarrec** [comment]
Ok I see where the issue is coming from, the `PlantEquipmentOperationSchemes` in the .idd only accepts a subset of schemes. The OutdoorDewpointDifference for eg is only accepted for `ConderserEquipmentOperationSchemes`.

```
PlantEquipmentOperationSchemes,
       \memo Operation schemes are listed in "priority" order.  Note that each scheme
       \memo must address the entire load and/or condition ranges for the simulation.
       \memo The actual one selected for use will be the first that is "Scheduled"
       \memo on.  That is, if control scheme 1 is not "on" and control scheme 2
       \memo is -- then control scheme 2 is selected.
       \memo Only plant equipment should be listed on a Control Scheme for this item.
       \min-fields 4
  A1 , \field Name
        \required-field
       \reference PlantOperationSchemes
  A2 , \field Control Scheme 1 Object Type
        \required-field
       \type choice
       \key PlantEquipmentOperation:CoolingLoad
       \key PlantEquipmentOperation:HeatingLoad
       \key PlantEquipmentOperation:Uncontrolled
       \key PlantEquipmentOperation:ComponentSetpoint
       \key PlantEquipmentOperation:ThermalEnergyStorage
       \key PlantEquipmentOperation:UserDefined
  A3 , \field Control Scheme 1 Name
```

Is that really waranted though?

**Comment by jmarrec** [comment]
Seems like we posted at the same exact time @mjwitte. 

As far as I'm concerned, I don't see a reason why a plant loop couldn't have one of these.

For example, I might have very good reasons to want to define a `PlantEquipmentOperation:OutdoorDryBulb` for a heating loop. I might want to allow a backup boiler to turn on only when it's really cold outside.

**Comment by jmarrec** [comment]
(More generally speaking, I'm not even sure why there's a need for `CondenserLoop` on top of `PlantLoop` anymore)

**Comment by Myoldmopar** [comment]
In August of 2009, myself along with several other developers worked to successfully merge the condenser loop logic into the plant manager, basically eliminating any difference between the two underneath the hood.  There were still gotchas but overall it went smoothly.  Because of the drastic changes behind the scenes, the input objects were left as separate citizens, to avoid any major hiccups for users.  

Then nearly a decade passed.  There was always a goal to eliminate the condenser loop, as it is indeed redundant, but it's hard to get a project just to simply eliminate an object.  

There's really nothing we can do at the moment to "fix" this.  As @mjwitte stated, this file simply shouldn't have run in 8.8 or 8.9.  To address this, I think we should probably eliminate the condenser loop now as part of a large bug fix.  We'll need to set up a broad set of testing, but it will be a great improvement on the input side.  @mjwitte I'll self assign here and work on this after I catch up on things from after the recent chaos.

**Comment by jmarrec** [comment]
I’d guess that openstudio caught up on the fact that condenserloop was redundant and to be deprecated (@kbenne’s flair probably!), and as a result it currently relies on PlantLoop for everything including loops with condensing towers, so this “regression” (as in “used to run but doesn’t now”) appears to be a major one as far as openstudio is concerned, in my opinion that is. Granted the majority of models don’t use these plant equipment objects, but still...

I’ll let smarter, more informed people take a stand, but myself, theoretically, I’d start by asking how soon there could be a subsequent e+ release where it’d work to gauge whether it’s even worth it to implement a **temporary** fix on openstudio‘s side. Just my two cents.

——
Sent from a mobile device, please excuse the brevity.
Julien Marrec, EBCP, BPI MFBA
Owner


Direct: +33 6 95 14 42 13
Website: www.effibem.com

LinkedIn (en) | (fr)

**Comment by mjwitte** [comment]
@jmarrec @kbenne I have confirmed locally that adding the missing keys to the idd for PlantEquipmentOperationSchemes and then re-building EnergyPlus (which rebuilds the internal JSON schema) fixes this.

**Comment by Myoldmopar** [comment]
Not sure what's going to happen on this, but here is a branch off of v8.9.0 (before style) with the IDD fixes pushed:  https://github.com/NREL/EnergyPlus/pull/6633

**Comment by Myoldmopar** [comment]
This was fixed up for 9.0 in #6633.
